src/recommendation/src/main/scala/RankingTrainValidationSplit.scala

```diff
@@ -0,0 +1,343 @@
+
+      val testds = dataset
+        .groupBy(col($(userCol)))
+        .agg(collect_list(col($(itemCol))))
+        .withColumn("shuffle", shuffle(col("collect_list(" + $(itemCol) + ")")))
+        .withColumn("train", sliceudf(col("shuffle")))
+        .withColumn("test", dropudf(col("shuffle")))
+        .drop(col("collect_list(" + $(itemCol) + ")")).drop(col("shuffle"))
+        .cache()
+
+      val train = testds
+        .select($(userCol), "train")
+        .withColumn($(itemCol), explode(col("train")))
+        .drop("train")
+
+      val test = testds
+        .select($(userCol), "test")
+        .withColumn($(itemCol), explode(col("test")))
+        .drop("test")
+
+      Array(train, test)
+    }
+  }
+
+  def prepareTestData(validationDataset: DataFrame, recs: DataFrame, k: Int): Dataset[_] = {
+    val est = $(estimator) match {
+      case p: Pipeline =>
+        //Assume Rec is last stage of pipeline
+        val pipe = $(estimator).asInstanceOf[Pipeline].getStages.last
+        pipe match {
+          case a: ALS =>
+            pipe.asInstanceOf[ALS]
+        }
+      case a: ALS =>
+        $(estimator).asInstanceOf[ALS]
+    }
+
+    val userColumn = est.getUserCol
+    val itemColumn = est.getItemCol
+
+    val perUserRecommendedItemsDF: DataFrame = recs
+      .select(userColumn, "recommendations." + itemColumn)
+      .withColumnRenamed(itemColumn, "prediction")
+
+    val perUserActualItemsDF = if (validationDataset.columns.contains($(ratingCol))) {
+      val windowSpec = Window.partitionBy(userColumn).orderBy(col($(ratingCol)).desc)
+
+      validationDataset
+        .select(userColumn, itemColumn, $(ratingCol))
+        .withColumn("rank", r().over(windowSpec).alias("rank"))
+        .where(col("rank") <= k)
+        .groupBy(userColumn)
+        .agg(col(userColumn), collect_list(col(itemColumn)))
+        .withColumnRenamed("collect_list(" + itemColumn + ")", "label")
+        .select(userColumn, "label")
+    } else {
+      val windowSpec = Window.partitionBy(userColumn).orderBy(col($(itemCol)).desc)
+
+      validationDataset
+        .select(userColumn, itemColumn)
+        .withColumn("rank", r().over(windowSpec).alias("rank"))
+        .where(col("rank") <= k)
+        .groupBy(userColumn)
+        .agg(col(userColumn), collect_list(col(itemColumn)))
+        .withColumnRenamed("collect_list(" + itemColumn + ")", "label")
+        .select(userColumn, "label")
+    }
+    val joined_rec_actual = perUserRecommendedItemsDF
+      .join(perUserActualItemsDF, userColumn)
+      .drop(userColumn)
+
+    joined_rec_actual
+  }
+}
+
+object RankingTrainValidationSplit extends ComplexParamsReadable[RankingTrainValidationSplit] {
+  def popRow(r: Row): Any = r.getDouble(1)
+}
+
+@InternalWrapper
+class RankingTrainValidationSplitModel(
+  override val uid: String)
+  extends Model[RankingTrainValidationSplitModel] with RankingTrainValidationSplitParams
```

Comment:
I don't think this is using all of those params right?  Those params are only on the estimator.  I don't think you need to include them on the model.  Also, since they are only on the estimator, you can move the set* methods to the params file next to the getters.